### ISSUE TITLE
Training mode by default?

### ISSUE THREAD

**Issue by Saulzar** [torch/torch7#707]
I've been running into a subtle issue with BatchNormalization, namely the fact that it's in training mode by default. 

Here's what I've been doing - and I've seen it in other places, I've been passing dummy data through the network in order to compute the sizes at each layer (because I'm lazy), usually an uninitialized Tensor with batch size of 1. Now if BatchNormalization is involved in the network, it 'learns' from that uninitialised data, which subsequently causes the network occasionally to fail to train properly. 

At least, that is what I think is happening, because if I'm careful to always use evaluation mode for that purpose, all is well. So, for that reason I think evaluation mode should be default.
